### PR TITLE
Adds Auth Result to Sign In with Custom token

### PR DESCRIPTION
**Pull request by protocol86**

### CODE REVIEW

File: Example/Auth/Sample/MainViewController.m [hunk 1]
```diff
@@ -1754,6 +1762,24 @@ - (void)signInWithCustomToken {
...
+ */
+- (void)signInWithCustomTokenAuthResult {
+  CustomTokenDataEntryViewControllerCompletion action =
+      ^(BOOL cancelled, NSString *_Nullable userEnteredTokenText) {
+        if (cancelled) {
+          [self log:@"CANCELLED:sign-in with custom token cancelled."];
+          return;
+        }
+
+        [self doSignInAndRetrieveDataWithCustomToken:userEnteredTokenText];
```
Comment (yramanchuk) [inline Example/Auth/Sample/MainViewController.m]: Should we use weak ref to self?
The same as 'StaticContentTableViewCell'

File: Firebase/Auth/Source/FIRAuth.m [hunk 2]
```diff
@@ -790,25 +790,23 @@ - (void)signInWithCustomToken:(NSString *)token
...
-        return;
-      }
-      [self completeSignInWithAccessToken:response.IDToken
-                accessTokenExpirationDate:response.approximateExpirationDate
-                             refreshToken:response.refreshToken
-                                anonymous:NO
-                                 callback:decoratedCallback];
+    [self internalSignInAndRetrieveDataWithCustomToken:token
+                                            completion:^(FIRAuthDataResult *_Nullable authResult,
+                                                         NSError * _Nullable error) {
```
Comment (yramanchuk) [inline Firebase/Auth/Source/FIRAuth.m]: nit: remove space '*_Nullable'

File: Firebase/Auth/Source/FIRAuth.m [hunk 3]
```diff
@@ -1196,6 +1194,48 @@ - (void)phoneNumberSignInWithRequest:(FIRVerifyPhoneNumberRequest *)request
 }
 #endif
 
+/** @fn internalSignInAndRetrieveDataWithCustomToken:completion:
+    @brief Signs in a Firebase user given a custom token.
+    @param completion A block which is invoked when the custom token sign in request completes.
```
Comment (yramanchuk) [inline Firebase/Auth/Source/FIRAuth.m]: Please add: 'param token ...'

File: Firebase/Auth/Source/FIRAuth.m [hunk 4]
```diff
@@ -1196,6 +1194,48 @@ - (void)phoneNumberSignInWithRequest:(FIRVerifyPhoneNumberRequest *)request
...
+- (void)internalSignInAndRetrieveDataWithCustomToken:(NSString *)token
+                           completion:(nullable FIRAuthDataResultCallback)completion {
+  FIRVerifyCustomTokenRequest *request =
+    [[FIRVerifyCustomTokenRequest alloc] initWithToken:token
+                                  requestConfiguration:_requestConfiguration];
+  [FIRAuthBackend verifyCustomToken:request
+                           callback:^(FIRVerifyCustomTokenResponse *_Nullable response,
+                                      NSError *_Nullable error) {
+    if (error) {
+      completion(nil, error);
```
Comment (yramanchuk) [inline Firebase/Auth/Source/FIRAuth.m]: completion block is nullable, please verify it's not null

File: Firebase/Auth/Source/FIRAuth.m [hunk 5]
```diff
@@ -1196,6 +1194,48 @@ - (void)phoneNumberSignInWithRequest:(FIRVerifyPhoneNumberRequest *)request
...
+      return;
+    }
+    [self completeSignInWithAccessToken:response.IDToken
+              accessTokenExpirationDate:response.approximateExpirationDate
+                           refreshToken:response.refreshToken
+                              anonymous:NO
+                               callback:^(FIRUser *_Nullable user,
+                                          NSError *_Nullable error) {
+      if (error){
```
Comment (yramanchuk) [inline Firebase/Auth/Source/FIRAuth.m]: nit: add space before '{'

File: Firebase/Auth/Source/FIRAuth.m [hunk 6]
```diff
@@ -1196,6 +1194,48 @@ - (void)phoneNumberSignInWithRequest:(FIRVerifyPhoneNumberRequest *)request
 }
 #endif
 
+/** @fn internalSignInAndRetrieveDataWithCustomToken:completion:
+    @brief Signs in a Firebase user given a custom token.
+    @param completion A block which is invoked when the custom token sign in request completes.
+ */
+- (void)internalSignInAndRetrieveDataWithCustomToken:(NSString *)token
+                           completion:(nullable FIRAuthDataResultCallback)completion {
```
Comment (yramanchuk) [inline Firebase/Auth/Source/FIRAuth.m]: nit: align colons 'completion:'

File: Firebase/Auth/Source/FIRAuth.m [hunk 7]
```diff
@@ -790,25 +790,23 @@ - (void)signInWithCustomToken:(NSString *)token
...
     }];
   });
 }
 
+- (void)signInAndRetrieveDataWithCustomToken:(NSString *)token
+                                  completion:(FIRAuthDataResultCallback)completion {
+  dispatch_async(FIRAuthGlobalWorkQueue(), ^{
+    FIRAuthDataResultCallback decoratedCallback =
+        [self signInFlowAuthDataResultCallbackByDecoratingCallback:completion];
+    [self internalSignInAndRetrieveDataWithCustomToken:token completion:decoratedCallback];
```
Comment (yramanchuk) [inline Firebase/Auth/Source/FIRAuth.m]: Just double checking: why this method has non-nullable completion param and 'internalSignInAndRetrieveDataWithCustomToken:completion:' has nullable 'completion'?

BTW: 'signInWithCustomToken: completion:' has nullable param as well.

File: Example/Auth/Tests/FIRAuthTests.m [hunk 8]
```diff
@@ -1248,6 +1248,60 @@ - (void)testSignInWithCustomTokenFailure {
...
+      id mockVeriyCustomTokenResponse = OCMClassMock([FIRVerifyCustomTokenResponse class]);
+      [self stubTokensWithMockResponse:mockVeriyCustomTokenResponse];
+      callback(mockVeriyCustomTokenResponse, nil);
+    });
+  });
+  [self expectGetAccountInfo];
+  XCTestExpectation *expectation = [self expectationWithDescription:@"callback"];
+  [[FIRAuth auth] signOut:NULL];
+  [[FIRAuth auth] signInAndRetrieveDataWithCustomToken:kCustomToken
+                                            completion:^(FIRAuthDataResult *_Nullable result,
```
Comment (yramanchuk) [inline Example/Auth/Tests/FIRAuthTests.m]: Should we verify result.additionalUserInfo.isNewUser property value?
We may add this for all places where this property is returned.

Comment (protocol86) [reply Example/Auth/Tests/FIRAuthTests.m]: Done, will update other places in a separate PR.

File: Firebase/Auth/Source/FIRAuth.m (hunk 6 above)
Comment (protocol86) [reply Firebase/Auth/Source/FIRAuth.m]: done

File: Firebase/Auth/Source/FIRAuth.m (hunk 4 above)
Comment (protocol86) [reply Firebase/Auth/Source/FIRAuth.m]: done.

File: Firebase/Auth/Source/FIRAuth.m (hunk 5 above)
Comment (protocol86) [reply Firebase/Auth/Source/FIRAuth.m]: Done.

File: Firebase/Auth/Source/FIRAuth.m (hunk 3 above)
Comment (protocol86) [reply Firebase/Auth/Source/FIRAuth.m]: done

File: Firebase/Auth/Source/FIRAuth.m (hunk 7 above)
Comment (protocol86) [reply Firebase/Auth/Source/FIRAuth.m]: In the header file this completion is nullable, the compiler somehow didn't complain about it being non-null here.  Done.

File: Firebase/Auth/Source/FIRAuth.m (hunk 2 above)
Comment (protocol86) [reply Firebase/Auth/Source/FIRAuth.m]: Done.

File: Example/Auth/Sample/MainViewController.m (hunk 1 above)
Comment (protocol86) [reply Example/Auth/Sample/MainViewController.m]: This is fine this way, weakself wouldn't impact it.

**Pull request merged**